File: Projeto_8/Atv - POO8/Projeto_8/Persistencia.cpp
#include "Persistencia.h"
#include <fstream>
#include <iostream>

void Persistencia::salvarDados(const std::vector<Aluno>& alunos, const std::vector<Professor>& professores, const std::vector<Disciplina>& disciplinas) {
    std::ofstream arquivo("dados.txt");
    if (!arquivo) throw std::runtime_error("Erro ao abrir o arquivo para escrita.");
    
    arquivo << alunos.size() << "\n";
    for (const auto& aluno : alunos) {
        arquivo << aluno.getNome() << " " << aluno.getIdade() << "\n";
    }
    
    arquivo << professores.size() << "\n";
    for (const auto& professor : professores) {
        arquivo << professor.getNome() << " " << professor.getEspecializacao() << "\n";
    }
    
    arquivo << disciplinas.size() << "\n";
    for (const auto& disciplina : disciplinas) {
        arquivo << disciplina.getNome() << " " << disciplina.getCargaHoraria() << "\n";
    }
}

void Persistencia::carregarDados(std::vector<Aluno>& alunos, std::vector<Professor>& professores, std::vector<Disciplina>& disciplinas) {
    std::ifstream arquivo("dados.txt");
    if (!arquivo) return; // Se não existe, inicia com lista vazia
    
    size_t tamanho;
    alunos.clear();
    professores.clear();
    disciplinas.clear();
    
    arquivo >> tamanho;
    for (size_t i = 0; i < tamanho; ++i) {
        std::string nome;
        int idade;
        arquivo >> nome >> idade;
        alunos.emplace_back(nome, idade);
    }
    
    arquivo >> tamanho;
    for (size_t i = 0; i < tamanho; ++i) {
        std::string nome, especializacao;
        arquivo >> nome >> especializacao;
        professores.emplace_back(nome, especializacao);
    }
    
    arquivo >> tamanho;
    for (size_t i = 0; i < tamanho; ++i) {
        std::string nome;
        int cargaHoraria;
        arquivo >> nome >> cargaHoraria;
        disciplinas.emplace_back(nome, cargaHoraria);
    }
}

File: Projeto_8/Projeto8-GabrielAlves/src/persistencia.py
# persistencia.py

import csv
from modelos import Aluno, Professor, Disciplina
from excecoes import ArquivoNaoEncontradoError, PermissaoNegadaError, FalhaConversaoDadosError

# Dicionário global para manter o estado dos dados em memória.
# Isso permite que o manipulador de sinal acesse os dados mais recentes.
dados_em_memoria = {
    "alunos": [],
    "professores": [],
    "disciplinas": []
}

def salvar_dados():
    """
    Salva todas as listas de dados (alunos, professores, disciplinas)
    em seus respectivos arquivos CSV.
    """
    print("\n[PERSISTÊNCIA] Tentando salvar dados...")
    try:
        # Salva Alunos
        with open('alunos.csv', 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(['id', 'nome'])
            for aluno in dados_em_memoria["alunos"]:
                writer.writerow([aluno.id, aluno.nome])

        # Salva Professores
        with open('professores.csv', 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(['id', 'nome'])
            for professor in dados_em_memoria["professores"]:
                writer.writerow([professor.id, professor.nome])

        # Salva Disciplinas
        with open('disciplinas.csv', 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(['id', 'nome'])
            for disciplina in dados_em_memoria["disciplinas"]:
                writer.writerow([disciplina.id, disciplina.nome])
        
        print("[PERSISTÊNCIA] Dados salvos com sucesso!")

    except PermissionError:
        # Trata erro de permissão negada ao gravar o arquivo [cite: 8]
        raise PermissaoNegadaError("alunos.csv, professores.csv ou disciplinas.csv")
    except Exception as e:
        print(f"Ocorreu um erro inesperado ao salvar os dados: {e}")

def carregar_dados():
    """
    Carrega os dados dos arquivos CSV para as listas em memória.
    Realiza o tratamento de exceções para arquivos não encontrados,
    permissões e erros de conversão de dados. 
    """
    print("[PERSISTÊNCIA] Carregando dados iniciais...")
    try:
        # Carrega Alunos
        with open('alunos.csv', 'r', encoding='utf-8') as f:
            reader = csv.reader(f)
            next(reader) # Pula o cabeçalho
            for linha in reader:
                if len(linha) != 2:
                    raise FalhaConversaoDadosError("Linha de aluno com formato inválido.") # [cite: 8]
                dados_em_memoria["alunos"].append(Aluno(int(linha[0]), linha[1]))

        # Carrega Professores
        with open('professores.csv', 'r', encoding='utf-8') as f:
            reader = csv.reader(f)
            next(reader) # Pula o cabeçalho
            for linha in reader:
                dados_em_memoria["professores"].append(Professor(int(linha[0]), linha[1]))

        # Carrega Disciplinas
        with open('disciplinas.csv', 'r', encoding='utf-8') as f:
            reader = csv.reader(f)
            next(reader) # Pula o cabeçalho
            for linha in reader:
                dados_em_memoria["disciplinas"].append(Disciplina(int(linha[0]), linha[1]))
        
        print("[PERSISTÊNCIA] Dados carregados com sucesso.")

    except FileNotFoundError as e:
        # Trata erro de arquivo não encontrado [cite: 8]
        print(f"[AVISO] Arquivo '{e.filename}' não encontrado. Iniciando com listas vazias.")
    except PermissionError as e:
        # Trata erro de permissão negada ao abrir o arquivo [cite: 8]
        raise PermissaoNegadaError(e.filename)
    except (ValueError, IndexError) as e:
        # Trata falhas na conversão de dados [cite: 8]
        raise FalhaConversaoDadosError(f"Não foi possível converter um valor no arquivo. Detalhes: {e}")
    except Exception as e:
        print(f"Ocorreu um erro inesperado ao carregar os dados: {e}")